.gitlab-ci.yml

```diff
@@ -2659,20 +2659,32 @@ latest_release_7:
     - if [[ -z "$TAG" ]]; then echo "Need a tag to delete"; exit 1; fi
     - inv -e docker.delete ${ORGANIZATION} ${IMAGE} ${TAG} ${DOCKER_TOKEN} &>/dev/null
 
-# invalidate cloudfront cache
-deploy_cloudfront_invalidate:
+#
+# Cloudfront cache invalidation:
+# Duplicated in 2 jobs: one that runs "on success" of the previous stage, and one that runs "on failure" of previous stages.
+# Compared to having 1 single job that runs "always", this setup guarantees that if earlier stages first failed and were
+# then retried successfully, the cloudfront invalidation will also run after the successful retry.
+#
+.deploy_cloudfront_invalidate: &deploy_cloudfront_invalidate
   <<: *run_when_triggered
   stage: deploy_invalidate
   image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/deploy:$DATADOG_AGENT_BUILDERS
   before_script:
     - ls $OMNIBUS_PACKAGE_DIR
   tags: [ "runner:main", "size:large" ]
-  when: always
   script:
     - cd /deploy_scripts/cloudfront-invalidation
     - "REPO=apt PATTERN_SUBSTRING=/$DEB_RPM_BUCKET_BRANCH/ ./invalidate.sh"
     - "REPO=yum PATTERN_SUBSTRING=/$DEB_RPM_BUCKET_BRANCH/ ./invalidate.sh"
 
+deploy_cloudfront_invalidate_on_success:
+  <<: *deploy_cloudfront_invalidate
+  when: on_success
+
+deploy_cloudfront_invalidate_on_failure:
+  <<: *deploy_cloudfront_invalidate
+  when: on_failure
```

Comment:
in general, I think it's safer to invalidate CloudFront when deploy jobs fail, as it guarantees that CloudFront accurately reflects what's in S3. 

For example:

* if one deploy job fails but not others, we still want to invalidate CloudFront
* if a deploy job fails in the middle/at the end of its run but is still able to update the repo in a consistent way, we still want to invalidate CloudFront 

The one case where invalidating CloudFront may not be desirable is when a deploy job literally breaks a staging repo, in which case we may want to count on the CloudFront cache to keep the previous state of the repo cached so that end users of the staging repo aren't affected. This could happen indeed, but even then not invalidating CloudFront may only add to the confusion (as CloudFront would not return the same objects as S3).